tensorflow/core/kernels/mkl/mkl_einsum_op.cc

```diff
@@ -0,0 +1,279 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#ifdef INTEL_MKL
+#define EIGEN_USE_THREADS
+#define EIGEN_DONT_PARALLELIZE
+#include "mkl_batch_matmul_helper.h"
+#include "tensorflow/core/kernels/linalg/einsum_op_impl.h"
+
+namespace tensorflow {
+
+typedef Eigen::ThreadPoolDevice CPUDevice;
+
+// MklEinsum Op is same as Eigen implementation of Einsum Op except,
+// it calls MKL's BatchMatMul implementation instead of Eigen implementation
+// for now, We are going to replace other Eigen Ops used here, with their
+// equivalent MKL ops for better performance.
```

Comment:
We should avoid having duplicate code. I normally don't want MKL code (`#ifdef INTEL_MKL`) interspersed with vanilla TF code, but I think I'm willing to trade that with having just one copy of code. Asking @ezhulenev for second opinion.

@ezhulenev This PR adds an MKL Einsum op. 

The op kernel is the same as the normal Einsum, except that it [calls MKLContractOperands](https://github.com/tensorflow/tensorflow/blob/8ab17489a4d4936404381b97b5c8135b7f20a9b3/tensorflow/core/kernels/mkl/mkl_einsum_op.cc#L182-L184) instead of [ContractOperands](https://github.com/tensorflow/tensorflow/blob/8ea9e8cfab457e73056347aa2366d748448a5745/tensorflow/core/kernels/linalg/einsum_op_impl.h#L639-L641). [MKLContractOperands](https://github.com/tensorflow/tensorflow/blob/8ab17489a4d4936404381b97b5c8135b7f20a9b3/tensorflow/core/kernels/mkl/mkl_einsum_op.cc#L39-L127) is also largely similar to [ContractOperands](https://github.com/tensorflow/tensorflow/blob/8ea9e8cfab457e73056347aa2366d748448a5745/tensorflow/core/kernels/linalg/einsum_op_impl.h#L546-L586), apart from the part that it calls DNNL batch matmul primitive.

I'm thinking we should just have one Einsum op, refactor the DNNL batch matmul part to a function, and call that function in `ContractOperands` under `#ifdef INTEL_MKL`. What do you think?

Reply:
I'd say it depends on the path of integrating mkl/oneDnn into TF that we'll choose. If we'll go with integrating MKL into "default" kernels, then it makes sense to provide different implementations of "helpers" and "functors" based on the compile flags (Eigen or MKL) and keep a single "Op" implementation.

Although current approach leads to a lot of code duplication I like that it's consistent with other MKL ops and that the code is completely separate.

I'd ask TF-Core owners what they think, maybe @rjpower can chime in.

Reply:
@rjpower  Any update on this PR? Please. Thanks!

Reply:
Having consistency with the other MKL ops seems appropriate, even if there's some extra duplication.

@penpornk / @ezhulenev if you feel it's appropriate, should we factor out the "helper" code for einsum into einsum-lib.h/cc and reuse it here?

Reply:
Thank you, Russell! Refactoring the helper code out sounds good to me. I want both ops to get the same fixes and optimizations.